### ISSUE TITLE
IO error: Failed to delete while purging index

### ISSUE THREAD

**Issue by mdoleary** [cockroachdb/cockroach#32593]
**Describe the problem**

We have a 3 node (version 2.1.0) cluster.  One of the nodes (n1) crashed running goroutine 97.  The only databases on the cluster are the system database and the StarTrek database.  No real activity yet.  The crash appeared to occur around 5:00 am during some maintenance routine.

**To Reproduce**

If possible, provide steps to reproduce the behavior:

1. Set up CockroachDB cluster (3 nodes.  2 Windows Server 2016 and 1 Debian 9 Linux all on same ESXi host).  The only databases in the cluster are the system database and the Startrek DB.

**Expected behavior**
Node should not crash when trying to purge an index.

**Additional data / screenshots**
The problem is not SQL-related.

If a node in your cluster encountered a fatal error, supply the contents of the
log directories (at minimum of the affected node(s), but preferably all nodes).

Note that log files can contain confidential information. Please continue
creating this issue, but contact support@cockroachlabs.com to submit the log
files in private.

```
C:\ustreem>cockroach start --certs-dir=certs --advertise-host=epu2016srv2 --join=epu2016srv1:26257,epud9srv1:26257
*
* WARNING: [n?] listen address "::" not in node certificate (IP=10.10.30.142; DNS=epu2016srv2; CN=node)
* Secure node-node and SQL connections are likely to fail.
* Consider extending the node certificate or tweak --listen-addr/--advertise-addr.
*
CockroachDB node starting at 2018-11-16 14:13:06.5341643 +0000 UTC (took 10.4s)
build:               CCL v2.1.0 @ 2018/10/30 12:40:27 (go1.10.3)
webui:               https://epu2016srv2:8080
sql:                 postgresql://root@epu2016srv2:26257?sslcert=certs%5Cclient.root.crt&sslkey=certs%5Cclient.root.key&sslmode=verify-full&sslrootcert=certs%5Cca.crt
client flags:        cockroach <client cmd> --host=epu2016srv2:26257 --certs-dir=certs
logs:                C:\ustreem\cockroach-data\logs
temp dir:            C:\ustreem\cockroach-data\cockroach-temp188306383
external I/O path:   C:\ustreem\cockroach-data\extern
store[0]:            path=C:\ustreem\cockroach-data
status:              restarted pre-existing node
clusterID:           223a9c54-4f04-46a0-8cb1-eff7b4f84c42
nodeID:              1
*
* ERROR: [n1,s1,r3/1:/System/NodeLiveness{-Max}] Reported as error b4ac1cb3e78542c5b3eb887375cd99df
*
F181119 11:32:36.364675 97 storage/replica.go:4511  [n1,s1,r3/1:/System/NodeLiveness{-Max}] while purging index %d: while purging index 580532: IO error: Failed to delete: C:\ustreem\cockroach-data\auxiliary\sideloading\3\3.1\i580532.t16: The system cannot find the path specified.
goroutine 97 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0xc042080d00, 0xc042080de0, 0x36f5900, 0x12)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:997 +0xd6
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x44f7c00, 0xc000000004, 0x36f59bc, 0x12, 0x119f, 0xc047eb03c0, 0xe7)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:864 +0x905
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x28a3140, 0xc04a66d170, 0x4, 0x2, 0x2348fbd, 0x6, 0xc04a2e5c00, 0x2, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:85 +0x2ec
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x28a3140, 0xc04a66d170, 0x1, 0x4, 0x2348fbd, 0x6, 0xc04a2e5c00, 0x2, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:69 +0x93
github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(0x28a3140, 0xc04a66d170, 0x2348fbd, 0x6, 0xc04a2e5c00, 0x2, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:172 +0x85
github.com/cockroachdb/cockroach/pkg/storage.fatalOnRaftReadyErr(0x28a3140, 0xc04a66d170, 0x236ec61, 0x16, 0x2878e00, 0xc04bb45420)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4511 +0x1cd
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue.func1(0x28a3140, 0xc04a66d170, 0xc042382e00, 0x28a3140)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3853 +0x17e
github.com/cockroachdb/cockroach/pkg/storage.(*Store).withReplicaForRequest(0xc0424ed080, 0x28a3140, 0xc04a66d170, 0xc04784f380, 0xc04a2e5ed0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3502 +0x13c
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc0424ed080, 0x28a3140, 0xc0429b2090, 0x3)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3840 +0x230
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc042786600, 0x28a3140, 0xc0429b2090)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:225 +0x222
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x28a3140, 0xc0429b2090)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc04232ea20, 0xc0421b34d0, 0xc04232ea10)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0xf0
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xb4


****************************************************************************

This node experienced a fatal error (printed above), and as a result the
process is terminating.

Fatal errors can occur due to faulty hardware (disks, memory, clocks) or a
problem in CockroachDB. With your help, the support team at Cockroach Labs
will try to determine the root cause, recommend next steps, and we can
improve CockroachDB based on your report.

Please submit a crash report by following the instructions here:

    https://github.com/cockroachdb/cockroach/issues/new/choose

If you would rather not post publicly, please contact us directly at:

    support@cockroachlabs.com

The Cockroach Labs team appreciates your feedback.

C:\ustreem>
```

**Environment:**
 - CockroachDB version 2.1.0
 - Server OS: Windows Server 2016
 - Client app: None

**Additional context**
The one node was marked as dead, but the other two nodes continued functioning.

**Comment by mdoleary** [comment]
Here are the logs from the affected server:

[epu2016srv2-logs.zip](https://github.com/cockroachdb/cockroach/files/2616027/epu2016srv2-logs.zip)

**Comment by mdoleary** [comment]
Accidentally closed the issue

**Comment by knz** [comment]
@tbg perhaps you could have a look? a crash in fatalOnRaftReadyErr seems to be up your alley

**Comment by knz** [comment]
@mdoleary what queries were being issued to the cluster at the time of the crash?

**Comment by mdoleary** [comment]
Hi Kena,

No queries were being issued to the cluster.  The cluster was created for evaluation purposes and there is no application traffic even going to it yet.

Regards,



Melodie 
O'Leary
Cleo  |  
Senior Test Engineer I
Tel: 815-282-7808
Email: moleary@cleo.com
  |  
Web: www.cleo.com
Harness the power of Cleo Integration Cloud to transform your business.

**Comment by tbg** [comment]
@mdoleary thanks for your report. I think this is a straightforward bug on our end, likely related to running on Windows.

(Technical details follow)

The error originates here. We expect file not found errors in this code path to occur, but they should be handled appropriately (and we test that, though I suspect not on windows).

https://github.com/cockroachdb/cockroach/blob/2087616a6718665c4a1074c412a6ea000952ea1b/pkg/storage/replica_sideload_disk.go#L117

The contract for `DeleteFile` states that `os.ErrNotExist` should be returned, but that's evidently not true because that would avoid the fatal crash.

https://github.com/cockroachdb/cockroach/blob/9a460556f58d1471ba4d0c7a1f6e092192b107e9/pkg/storage/engine/engine.go#L306-L308

The error returned is something like `IO error: Failed to delete:...` which is a RocksDB error.

This error isn't converted appropriately because of this janky code:

https://github.com/cockroachdb/cockroach/blob/c98e30b98f462c4a5c1c445f219d7841c5e4abd9/pkg/storage/engine/rocksdb.go#L2966-L2971

which doesn't match on Windows (it would have to match "The system cannot find the path specified.").

cc @benesch, are we running CI on Windows (I suspect we don't)? If so, I wonder why this test didn't fail:

https://github.com/cockroachdb/cockroach/blob/cbf2373b66ac781706ed4deb0e2290d1cf958583/pkg/storage/engine/rocksdb_test.go#L1204-L1225

**Comment by tbg** [comment]
@mdoleary this should be fixed in 2.1.2, slated for release on December 10th. Thanks again for letting us know about this problem!
